### ISSUE TITLE
Future of Watson: Community Collaboration?

### ISSUE THREAD

**Issue by dGilli** [TailorDev/Watson#504]
Hello everyone,

I've noticed there hasn't been any active development on Watson in a year, with v2 features hanging in the pipeline. Some of the planned features, like adding notes to time entries, could be game-changers for many of us.

It seems TailorDev no longer exists.

To the Maintainer(s):
Is there a possibility to give the community access or even transfer ownership, so we can take this project further?

I want to clarify that I have no intention of overtaking the project myself. I'm merely opening up this discussion as a potential entry point for the community to engage and possibly collaborate on the development to keep the project alive.

To the Community:
Would anyone else be interested in contributing if the project were opened up? Your thoughts and willingness to participate would be great to hear! Please comment.

I love what this tool has provided so far, and I believe some specific features while keeping it as simple as possible could add a lot of value.

**Comment by MaxG87** [comment]
I too would love to see this project being revived. I am not a power user, but even my simple use cases suffers some bugs. I certainly would contribute patches, but I am not sure if I could find the time to be a core maintainer.

**Comment by joelostblom** [comment]
I would love to see Watson remain an actively maintained project. There is a lot of fantastic work that has gone into it and to me it is the most convenient tool to keep track of time by a wide margin. I'm personally capable of contributing occasional patches, but not being a core dev.

@jmaupetit and @k4nar What are your thoughts on what the future of Watson looks like?

**Comment by k4nar** [comment]
I personally don't use Watson anymore, hence my lack of motivation to keep maintaining it :sweat_smile: . I would be happy to see other people taking the lead :+1: .

**Comment by ghost** [comment]
The watson feature set is quite complete. So I'm not sure we need a set of core devs to further build it out. Maybe a couple people maintinaing it and organizing issues and PRs is already good enough.

**Comment by nv1t** [comment]
I don't think it needs a core dev either, but i don't think it is feature complete 
Having somebody managing PRs and Maintaining the project would be nice, as there are PRs dangling for ever now.

I would love to help to maintain it, as i really use it productively.

**Comment by mdtrooper** [comment]
Well, this thread is one month more or less. What are the next steps?

- is it create a free organization into github?
- is it set members in these organization? 
- is it copy the git repo?
- is it posible move the issues and pull request in this github repo to another github repo?

**Comment by k4nar** [comment]
I can reach @jmaupetit if someone wants to be added as a owner of the project. But maybe the first step would be for this person(s) to go through the open PRs & issues, and see what we can merge?

**Comment by younes-io** [comment]
is this project maintained or abandoned ?
thank you

**Comment by uwe-schwarz** [comment]
I'm using this project quite extensively and would like to contribute. I would be willing to go through PRs and issues.
OTOH I must admit that I'm not a good coder, I mainly write some small scripts and use them all over the place. Therefore I can provide support and test PRs, merge them and make sure they are tested and work, but I wont write anything really new.

**Comment by s-frick** [comment]
I recently started using this project. I love it because of its simplicity. I'm not a python developer, but I have enough experience with many other languages that I should be able to pick it up quickly. I would be willing to maintain the project. @k4nar so feel free to have @jmaupetit add me as an owner if that's okay with you guys.

**Comment by jmaupetit** [comment]
As I've already told to some of you, and some others that contacted me directly, I've no time to maintain this project anymore. Even if I use it on a daily basis :sweat_smile: 

I propose to move it to the @jazzband organization and leave it to the community. Hope you'll all be ok with this.

:heart:

**Comment by mdtrooper** [comment]
> As I've already told to some of you, and some others that contacted me directly, I've no time to maintain this project anymore. Even if I use it on a daily basis 😅
> 
> I propose to move it to the @jazzband organization and leave it to the community. Hope you'll all be ok with this.
> 
> ❤️

I didn't know this community, but I am reading about they and they are awesome (I am in another similar community but it is about  abandoned floss games) .

I vote +1 .

One technical question: Can you move all "github things" (issues, pull request...) of this repo to jazzband?

**Comment by jmaupetit** [comment]
> One technical question: Can you move all "github things" (issues, pull request...) of this repo to jazzband?

Looks like it is :point_right:  https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#whats-transferred-with-a-repository

**Comment by timokau** [comment]
That sounds good, thank you for putting time into the handoff @jmaupetit!

One thing is not quite clear to me though: Are there any access controls in jazzband? If yes, how does that work? If not, does that mean that all 1000+ members would have commit access to watson?

**Comment by s-frick** [comment]
I very much welcome your suggestion. Organizations like @jazzband and their work are absolutely important. But I'm not quite sure if this is the right place for this project in particular. When I look at the repos, there are 90% django related libs. Watson is a cli tool that is useful to a wider audience regardless of the technological context and doesn't have much to do with python web frameworks. 
I think there are a few people here who could collaboratively maintain watson and that would be the preferred solution for me personally. 

But whatever you decide thank you for your efforts and the time you invest to keep this going.

**Comment by b12f** [comment]
I agree that the @jazzband org seems a bit more python infrastructury-oriented, and creating a separate org might be best. It's hard to figure out who might be good and trustworthy new stewards though, so I understand the idea of just moving it into the hands of an established community, where development can be restarted. It can then still be decided to spin the project off into a separate org.

:+1: for moving it into @jazzband, unless the current maintainers think they can find a group of like 3 peeps they trust that would manage an extra org.

**Comment by imhosseinzadeh** [comment]
I hope the development of this project will start again. I recently looked for a time logger and found Watson. there isn't any reliable alternative (CLI or GUI).

**Comment by lwjohnst86** [comment]
What's the status of this move to @jazzband? I'd love to see this continue too, it's the best time tracker that I've found (at least for my purposes).

**Comment by austinorth** [comment]
Just chiming in to say I love this tool for my own personal use. My company doesn't require any sort of time tracking, but I still use it to kind of reflect on my day and stuff. I think @b12f has a good point here. If @jazzband is down to host the repo for now, the community can always spin out to another organization at later date. 🤷🏻 @jmaupetit if you need someone to champion the migration over and such, I'd be down to join their matrix chat and get the ball rolling.

**Comment by oriongonza** [comment]
@austinorth Do you think you could do it?

**Comment by austinorth** [comment]
@dev-ardi hey yup I'm still definitely down to handle the migration! I'd just need permissions granted by and the blessing of the folks that are currently repo admins and whatnot (@jmaupetit @k4nar @SpotlightKid are the top 3 contributors, so tagging here for visibility).

**Comment by kuon** [comment]
I looked at command line time tracking tool, and I started using watson.

I already implemented some features and fixes for my use, but I didn't do any PR as this project seems idle at the moment.

As a long time open source contributor and maintainer, I think it is really important to be sure that the project can be managed by the community, and that nobody "hold the keys" and blocks everything. BTW: this is not a critic of the people starting the project and not having time or motivation to continue.

I don't know jazzband that much as I am not a big python user, but it looks like a good fit.

I hope you can migrate the project. But if it doesn't move in the coming weeks, I will open my fork and share my changes as I think it would help the community.

Cheers

**Comment by b12f** [comment]
Hello everyone, we've decided to host a repo that enables a restart of development at https://github.com/pub-solar/Watson

https://pub.solar is a small tech community mostly focused on hosting a FOSS suite of applications, but we also organize a hackathon every once in a while. We're currently 4 administrative people, with a couple more in the community.

We probably won't be making too many changes to the codebase of Watson ourselves, but can manage and review PRs. Most of us are active users of Watson. Should @jazzband jump in and also take the project, we're more than happy to hand it over.

**Comment by bjorn** [comment]
@b12f It's nice that you decided to pick up the project, but why is your repository not a proper fork of this one? This Watson repository has a network of 236 forks attached to it and your repository is entirely disconnected.

I think apart from it just being good style and giving attribution, staying within the same network also makes it possible for pull requests to be opened from any of the existing forks, whereas now people would have to fork again.

**Comment by teutat3s** [comment]
@bjorn We added a hint to the README and short description on the right side menu and all commit history is there for proper attribution. The problem with GitHub's fork UX is the following:
1. Fork a repo
2. Be a happy forker with an active fork
3. A second forker appears and forks the fork
4. 2nd forker makes changes and would like to contribute to the first fork
5. 2nd forker quickly opens a PR "click", "click", "click" without looking too much at the details
6. GitHub will happily create the PR against the ORIGINAL repo instead of the first fork if you don't watch out
7. Sad 2nd forker closes the PR against the inactive ORIGINAL repo and opens another PR against first fork

Hopefully I was able to explain why we did not use GitHub's fork feature.

*Disclosure:* parts of this UX story might have been experienced by the author of this comment.

**Comment by kuon** [comment]
Yeah, "hard forking" is the way to go because of many little github UX quirk. Also if a repo is closed under certain circumstances, the forks can be inaccessible too.

**Comment by bjorn** [comment]
I think the solution in that case is to open an issue with GitHub support. It is also possible to change the parent repository that way (so the owner of this repo could point it to the new maintained one, for example).

**Comment by oriongonza** [comment]
The solution is what pub.solar did. It's properly attributed in the README, I don't see how this is an issue.

I also don't like seeing the `Forked from...` because to me that implies that it's a fork that actively contributes to upstream, and this fork _is_ upstream.

**Comment by bjorn** [comment]
> I also don't like seeing the `Forked from...` because to me that implies that it's a fork that actively contributes to upstream, and this fork _is_ upstream.

I think this is a problem with interpretation and it's also weird to me that the built-in "Forked from ..." would be problematic whereas a manual "Community fork of ..." in the description is somehow better.

The owners of this repository could just move the repository to its new location, which would have the advantage of keeping existing PRs, issues, stars and forks, as well as leaving a redirect in place. If they still want a copy at @TailorDev then they can fork it back. As far as I know, GitHub support can help change the upstream repository if needed.

Anyway, I have to say I'm rather surprised by the apparent general agreement that creating a new bare repository is the way forward here. I have personally handled similar maintainership changes by just moving repositories or having the previous maintainer do this. I didn't intend for this to turn into a big discussion.

**Comment by vladh** [comment]
Hey @jmaupetit, shall we transfer this to @jazzband now, so that issues etc. are preserved?

If not, could you let us know so that we can decide on a fork as canonical? Thank you!

**Comment by b12f** [comment]
I also want to chime in and say that Githubs forking UX is bad because it will always show the original repository as somehow a "parent" or "upstream" of the repo you're looking at. I don't think this makes any sense at all for projects where the original repository has been abandoned or where the fork has become a different project. Imagine if the mariadb repo would still be linking back to mysql as if the mysql repo was somehow authoritative.

I've even thought about forking to our own git server at git.pub.solar. Git is by design a distributed system, and anyone being mad at the way we've forked is a bit too used to the Github way or doing things imo. We could also go the sr.ht route and only accept patches via email. In any case, we'll be doing on-and-off development and keeping the package alive for our usecases over in our repository, and anyone that wants to join in is free to do so.

**Comment by vladh** [comment]
Thanks @b12f. I’ve directly emailed @jmaupetit in a last attempt to migrate this repository wholesale, preserving issues etc. In my opinion, if that’s not successful, considering a hard fork as canonical is the only way.

**Comment by acidjunk** [comment]
This process starts to feel messy. It would be nice to have one obvious path forward. (except if your Dutch)

Relying on several community members with one or more core maintainers, would be my first preference. (backed by a foundation?)

Jazzband would be my second preference: because it seems reasonable to expect that issues and open PR's can be moved.

Other forking ways could also work; but would have the downside that the pypi package will be renamed. Which will be less favorable from a users perspective.

This would be a good time to act.

**Comment by vladh** [comment]
@acidjunk I agree. The only way to have a clean continuation of maintainership is to get access from the original maintainer. Let’s wait for a reply for a moment longer. If we don’t have that, we can decide on a single hard fork going forward.

**Comment by bjorn** [comment]
> I don't think this makes any sense at all for projects where the original repository has been abandoned or where the fork has become a different project. Imagine if the mariadb repo would still be linking back to mysql as if the mysql repo was somehow authoritative.

Yes, in that case it makes no sense and one would just create a detached fork (what used to be the original meaning of "fork", before GitHub came along and made forks into a way of collaboration - such that we now need to talk about "hard forks"). But it's irrelevant here because we're not looking to split off the project but just to find a new maintainer.

If the new maintainer wants to host on a different platform then they can handle the migration any way they want, but if they are also on GitHub then the best approach is just to move the repository.

**Comment by jmaupetit** [comment]
I'll move the repository to @jazzband later this week. I will still be in charge of releases publication to PyPI (until a clear maintainers board is constituted).

**Comment by vladh** [comment]
Let's make it clear who's willing to commit to being a maintainer. It looks like so far we have myself, @b12f, and @teutat3s. Do I have that right? 3 people seems enough to me.

**Comment by jmaupetit** [comment]
We moved to @jazzband! Still some work to do :sweat_smile: :point_right: #509

**Comment by vladh** [comment]
Amazing! Thanks so much @jmaupetit!

**Comment by loeffel-io** [comment]
I use watson for auto time tracking in tmux - its amazing! please let watson not die ❤️
